File: Average/Javascript/jcla1/average.js
function arithmetic_mean(arr) {
    return sum(arr) / arr.length;
}

function geometric_mean(arr) {
    return Math.pow(prod(arr), 1 / arr.length);
}

function harmonic_mean(arr) {
    return arr.length / sum(arr.map(function(n) {
        return 1 / n;
    }));
}

function generalized_mean(arr, power) {
    return Math.pow((1 / arr.length) * sum(arr.map(function(n) {
        return Math.pow(n, power);
    })), 1 / power);
}

function weighted_mean(arr, weights) {
    if (arr.length != weights.length) throw "# values & # weights have to be the same!";

    var weighted = 0;
    for (var i = arr.length - 1; i >= 0; i--) {
        weighted += arr[i] * weights[i]
    }

    return weighted / sum(weights);
}

function midrange_mean(arr) {
    return 0.5 * (max(arr) + min(arr))
}

function energatic_mean(arr) {
    return 10 * log(sum(arr.map(function(n) {
        return Math.pow(10, n / 10)
    })) / arr.length, 10);
}

function log(a, b) {
    return Math.log(b) / Math.log(a);
}

function max(arr) {
    return arr.reduce(function(acc, n) {
        if (n > acc) return n;
        return acc;
    })
}

function min(arr) {
    return arr.reduce(function(acc, n) {
        if (n < acc) return n;
        return acc;
    })
}


function sum(arr) {
    return arr.reduce(function(acc, n) {
        return acc + n;
    });
}

function prod(arr) {
    return arr.reduce(function(acc, n) {
        return acc * n;
    });
}

File: Average/Scala/jcla1/average.scala
object Average {
    def arithmeticMean(xs: List[Double]): Double = xs.sum / xs.size

    def geometricMean(xs: List[Double]): Double =
        math.pow(xs.product, 1.0 / xs.size)

    def harmonicMean(xs: List[Double]): Double =
        xs.size  / xs.map(n => 1.0 / n).sum

    def generalizedMean(xs: List[Double], power: Double): Double =
        math.pow(1.0 / xs.size * xs.map(n => math.pow(n, power)).sum,
            1.0 / power)

    def weightedMean(xs: List[Double], weights: List[Double]): Double =
        xs.zip(weights).map(t => t._1 * t._2).sum / weights.sum

    def midrangeMean(xs: List[Double]): Double =
        0.5 * (xs.max + xs.min)

    def energaticMean(xs: List[Double]): Double =
        10 * math.log10(xs.map(n => math.pow(n, n / 10)).sum / xs.size)
}